File: typescript/packages/core/src/evals/case-dir.ts
import { existsSync, readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { z } from "zod";
import { KnownEvent } from "../log";
import type { Lit, Opt, Strict } from "../log/zod-types";
import { scriptProblem } from "../model/scripted";
import { err, ok, type Result } from "../result";
import { SandboxScript } from "../sandbox/script";
import type { EventDraft } from "../store";
import {
  CaseLine0,
  CaseOffline,
  CaseSnapshot,
  ExtensionScript,
  SandboxResults,
} from "./files";
import { EventMatcher, Rubric } from "./schema";

// One saved case directory, read and parsed at the boundary (spec/conformance/case.schema.json,
// spec/schema/eval.v1.schema.json). The case's files are user input: anything malformed makes
// the case an `error`, never a throw.

const IMPL = "threads-ts";

type Matchers = z.ZodArray<typeof EventMatcher>;
const CaseFile: Strict<{
  name: z.ZodString;
  family: z.ZodString;
  kind: Lit<"stub">;
  description: z.ZodString;
  clock: Strict<{ now: z.ZodInt }>;
  model_script: Opt<Lit<"model.json">>;
  sandbox_script: Opt<Lit<"sandbox.json">>;
  stub_script: Opt<Lit<"stubs.json">>;
  extension_script: Opt<Lit<"extensions.json">>;
  input: Opt<Strict<{ text: Opt<z.ZodString> }>>;
  expect: Strict<{ must: Matchers; expect: Opt<Matchers> }>;
  rubric: Opt<typeof Rubric>;
  snapshot: Opt<typeof CaseSnapshot>;
  offline: Opt<typeof CaseOffline>;
  line0: Opt<typeof CaseLine0>;
}> = z.strictObject({
  name: z.string(),
  family: z.string(),
  kind: z.literal("stub"),
  description: z.string(),
  clock: z.strictObject({ now: z.int().min(0) }),
  model_script: z.literal("model.json").optional(),
  sandbox_script: z.literal("sandbox.json").optional(),
  stub_script: z.literal("stubs.json").optional(),
  extension_script: z.literal("extensions.json").optional(),
  input: z.strictObject({ text: z.string().optional() }).optional(),
  expect: z.strictObject({
    must: z.array(EventMatcher).min(1),
    expect: z.array(EventMatcher).optional(),
  }),
  rubric: Rubric.optional(),
  snapshot: CaseSnapshot.optional(),
  offline: CaseOffline.optional(),
  line0: CaseLine0.optional(),
});
export type CaseFile = z.infer<typeof CaseFile>;

/** expected.<impl>.json as the runner reads it: the recorded turn, or old-format matchers. */
const Expected = z.looseObject({
  appended: z.array(z.unknown()).optional(),
});

export type CaseDir = {
  readonly name: string;
  readonly meta: CaseFile;
  readonly log: Uint8Array;
  readonly artifacts: readonly Uint8Array[];
  readonly model: unknown;
  readonly sandbox: SandboxResults | SandboxScript | undefined;
  readonly stubs: unknown;
  readonly extensions: ExtensionScript | undefined;
  readonly line0: Uint8Array | undefined;
  /** The recorded turn, in full; undefined for a case saved before full events were kept. */
  readonly recorded: readonly KnownEvent[] | undefined;
  /**
   * An older case's `appended` matchers; undefined when it has none (a Python save_case before
   * this lane wrote only the outcome and state), so only its `must` and scripts are checked.
   */
  readonly matchers: readonly z.infer<typeof EventMatcher>[] | undefined;
};

type Read = Result<unknown, string>;

function json(dir: string, file: string): Read {
  const path = join(dir, file);
  if (!existsSync(path)) return err(`${file} is missing`);
  try {
    return ok(JSON.parse(readFileSync(path, "utf8")));
  } catch (error) {
    return err(
      `${file}: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
}

function parsed<T>(
  read: Read,
  schema: z.ZodType<T>,
  file: string,
): Result<T, string> {
  if (!read.ok) return read;
  const got = schema.safeParse(read.value);
  return got.success
    ? ok(got.data)
    : err(`${file}: ${got.error.issues[0]?.message ?? "invalid"}`);
}

/** sandbox.json: v2 results, else the v1 shape (one entry per tool name). */
function sandboxOf(
  dir: string,
  meta: CaseFile,
): Result<CaseDir["sandbox"], string> {
  if (meta.sandbox_script === undefined) return ok(undefined);
  const read = json(dir, "sandbox.json");
  if (!read.ok) return read;
  const v2 = SandboxResults.safeParse(read.value);
  if (v2.success) return ok(v2.data);
  return parsed(read, SandboxScript, "sandbox.json");
}

function bytes(path: string): Uint8Array | undefined {
  return existsSync(path) ? new Uint8Array(readFileSync(path)) : undefined;
}

/** The recorded turn: full events, or the matchers an older saveCase wrote. */
function appendedOf(
  read: Read,
): Result<Pick<CaseDir, "recorded" | "matchers">, string> {
  const expected = parsed(read, Expected, "expected.json");
  if (!expected.ok) return expected;
  const items = expected.value.appended;
  if (items === undefined)
    return ok({ recorded: undefined, matchers: undefined });
  const events = z.array(KnownEvent).safeParse(items);
  if (events.success && items.length > 0)
    return ok({ recorded: events.data, matchers: [] });
  const matchers = z.array(EventMatcher).safeParse(items);
  return matchers.success
    ? ok({ recorded: undefined, matchers: matchers.data })
    : err("expected.json: appended holds neither events nor matchers");
}

/** A file saveCase writes once per implementation: this one's name, else the shared one. */
function own(dir: string, name: string): string {
  const dot = name.indexOf(".");
  const mine = `${name.slice(0, dot)}.${IMPL}${name.slice(dot)}`;
  return existsSync(join(dir, mine)) ? mine : name;
}

export function readCase(root: string, name: string): Result<CaseDir, string> {
  const dir = join(root, name);
  const meta = parsed(json(dir, "case.json"), CaseFile, "case.json");
  if (!meta.ok) return meta;
  const log = bytes(join(dir, own(dir, "log.jsonl")));
  if (log === undefined) return err("log.jsonl is missing");
  const model =
    meta.value.model_script === undefined
      ? ok(undefined)
      : json(dir, "model.json");
  if (!model.ok) return model;
  const problem =
    model.value === undefined ? undefined : scriptProblem(model.value);
  if (problem !== undefined) return err(`model.json: ${problem}`);
  const sandbox = sandboxOf(dir, meta.value);
  if (!sandbox.ok) return sandbox;
  const stubs =
    meta.value.stub_script === undefined
      ? ok(undefined)
      : json(dir, "stubs.json");
  if (!stubs.ok) return stubs;
  const extensions =
    meta.value.extension_script === undefined
      ? ok(undefined)
      : parsed(
          json(dir, "extensions.json"),
          ExtensionScript,
          "extensions.json",
        );
  if (!extensions.ok) return extensions;
  const appended = appendedOf(json(dir, own(dir, "expected.json")));
  if (!appended.ok) return appended;
  const artifactDir = join(dir, "artifacts");
  return ok({
    name,
    meta: meta.value,
    log,
    artifacts: existsSync(artifactDir)
      ? readdirSync(artifactDir).map(
          (f) => new Uint8Array(readFileSync(join(artifactDir, f))),
        )
      : [],
    model: model.value,
    sandbox: sandbox.value,
    stubs: stubs.value,
    extensions: extensions.value,
    line0: bytes(join(dir, "line0.json")),
    ...appended.value,
  });
}

/** What the rerun sends: the recorded user_input as it was appended, or the case's text. */
export function inputOf(c: CaseDir): EventDraft | undefined {
  const recorded = c.recorded?.find((e) => e.type === "user_input");
  if (recorded?.type === "user_input") {
    const { type, type_version, critical, actor, data } = recorded;
    return { type, type_version, critical, actor, data };
  }
  const text = c.meta.input?.text;
  return text === undefined
    ? undefined
    : {
        type: "user_input",
        type_version: 1,
        critical: true,
        actor: {
          kind: "user",
          principal: { issuer: "api", tenant: "local", subject: "operator" },
        },
        data: { source: "api", text },
      };
}

/** Every case directory under `root`, by name. */
export function caseNames(root: string): readonly string[] {
  if (!existsSync(root)) return [];
  return readdirSync(root, { withFileTypes: true })
    .filter(
      (d) => d.isDirectory() && existsSync(join(root, d.name, "case.json")),
    )
    .map((d) => d.name)
    .toSorted();
}

File: python/src/threads/evals/case_dir.py
"""One saved case directory, read and parsed at the boundary (spec/conformance/case.schema.json,
spec/schema/eval.v1.schema.json). The case's files are user input: anything malformed makes the
case an `error`, never a raise."""

import json
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Annotated, Final, Literal

from pydantic import Field, JsonValue, StrictInt, StrictStr, TypeAdapter, ValidationError

from threads._generated.eval_v1 import (
    CaseLine0,
    CaseOffline,
    CaseSnapshot,
    EventMatcher,
    ExtensionScript,
    Rubric,
    SandboxResult,
    SandboxResults,
)
from threads._strict_model import StrictModel
from threads.evals.compare import Wire
from threads.log import Event
from threads.loop.scripted import scripted_model
from threads.reduce.handlers import to_json
from threads.result import Err, Ok

IMPL: Final = "threads-py"


class _Clock(StrictModel):
    now: Annotated[StrictInt, Field(ge=0)]


class _Input(StrictModel):
    text: StrictStr | None = None


class _Expect(StrictModel):
    must: Annotated[Sequence[EventMatcher], Field(min_length=1)]
    expect: Sequence[EventMatcher] = ()


class CaseFile(StrictModel):
    """case.json as the runner reads it: a saved (stub kind) case."""

    name: StrictStr
    family: StrictStr
    kind: Literal["stub"]
    description: StrictStr
    clock: _Clock
    model_script: Literal["model.json"] | None = None
    sandbox_script: Literal["sandbox.json"] | None = None
    stub_script: Literal["stubs.json"] | None = None
    extension_script: Literal["extensions.json"] | None = None
    input: _Input | None = None
    expect: _Expect
    rubric: Rubric | None = None
    snapshot: CaseSnapshot | None = None
    offline: CaseOffline | None = None
    line0: CaseLine0 | None = None


@dataclass(frozen=True, slots=True)
class CaseSandbox:
    """sandbox.json: v2 results, or the v1 shape (one entry per tool name)."""

    results: tuple[SandboxResult, ...] = ()
    v1: Mapping[str, Mapping[str, JsonValue]] | None = None


@dataclass(frozen=True, slots=True)
class CaseDir:
    name: str
    meta: CaseFile
    log: bytes
    artifacts: tuple[bytes, ...]
    model: JsonValue
    sandbox: CaseSandbox | None
    stubs: JsonValue
    extensions: ExtensionScript | None
    line0: bytes | None
    recorded: tuple[Event, ...] | None
    """The recorded turn, in full; None for a case saved before full events were kept."""
    matchers: tuple[Wire, ...] | None
    """An older case's `appended` matchers; None when its expected file has no `appended`."""

    @property
    def must(self) -> tuple[Wire, ...]:
        return tuple(_wire(m) for m in self.meta.expect.must)


type Read = Ok[JsonValue] | Err[str]

_JSON: Final = TypeAdapter[JsonValue](JsonValue)
_EVENTS: Final = TypeAdapter[list[Event]](list[Event])
_V1: Final = TypeAdapter[dict[str, dict[str, JsonValue]]](dict[str, dict[str, JsonValue]])


def _wire(m: EventMatcher) -> Wire:
    wire = to_json(m)
    return wire if isinstance(wire, dict) else {}


def _read(folder: Path, file: str) -> Read:
    path = folder / file
    if not path.exists():
        return Err(f"{file} is missing")
    try:
        return Ok(_JSON.validate_json(path.read_bytes()))
    except ValidationError as error:
        return Err(f"{file}: {error.errors()[0]['msg']}")


def _model[T: StrictModel](read: Read, model: type[T], file: str) -> Ok[T] | Err[str]:
    if isinstance(read, Err):
        return read
    try:
        return Ok(model.model_validate_json(json.dumps(read.value)))
    except ValidationError as error:
        return Err(f"{file}: {error.errors()[0]['msg']}")


def _sandbox(folder: Path, meta: CaseFile) -> Ok[CaseSandbox | None] | Err[str]:
    if meta.sandbox_script is None:
        return Ok(None)
    read = _read(folder, "sandbox.json")
    if isinstance(read, Err):
        return read
    v2 = _model(read, SandboxResults, "sandbox.json")
    if isinstance(v2, Ok):
        return Ok(CaseSandbox(results=tuple(v2.value.results)))
    tools = read.value.get("tools") if isinstance(read.value, dict) else None
    try:
        v1 = _V1.validate_python(tools, strict=True)
    except ValidationError:
        return Err("sandbox.json: neither v2 results nor a v1 tools map")
    if not all(isinstance(t.get("output"), str) for t in v1.values()):
        return Err("sandbox.json: a v1 tool needs an output string")
    return Ok(CaseSandbox(v1=v1))


def _appended(
    read: Read,
) -> Ok[tuple[tuple[Event, ...] | None, tuple[Wire, ...] | None]] | Err[str]:
    """The recorded turn: full events, or the matchers an older save_case wrote; neither for an
    old Python case, whose expected file keeps no `appended`."""
    if isinstance(read, Err):
        return read
    if isinstance(read.value, dict) and "appended" not in read.value:
        return Ok((None, None))
    items = read.value.get("appended", []) if isinstance(read.value, dict) else None
    if not isinstance(items, list):
        return Err("expected.json: appended is not a list")
    if items:
        try:
            return Ok((tuple(_EVENTS.validate_json(json.dumps(items))), ()))
        except ValidationError:
            pass
    matchers: list[Wire] = []
    for item in items:
        try:
            matchers.append(_wire(EventMatcher.model_validate_json(json.dumps(item))))
        except ValidationError:
            return Err("expected.json: appended holds neither events nor matchers")
    return Ok((None, tuple(matchers)))


def _own(folder: Path, name: str) -> str:
    """A file save_case writes once per implementation: this one's name, else the shared one."""
    stem, dot, ext = name.partition(".")
    mine = f"{stem}.{IMPL}{dot}{ext}"
    return mine if (folder / mine).exists() else name


def _optional(folder: Path, file: str | None) -> Read:
    return Ok(None) if file is None else _read(folder, file)


@dataclass(frozen=True, slots=True)
class _Scripts:
    model: JsonValue
    sandbox: CaseSandbox | None
    stubs: JsonValue
    extensions: ExtensionScript | None


def _model_script(folder: Path, meta: CaseFile) -> Read:
    model = _optional(folder, meta.model_script)
    if isinstance(model, Err) or model.value is None:
        return model
    if not isinstance(model.value, dict):
        return Err("model.json: not a model script")
    try:
        scripted_model(model.value)
    except (ValueError, ValidationError) as error:
        return Err(f"model.json: {error}")
    return model


def _scripts(folder: Path, meta: CaseFile) -> Ok[_Scripts] | Err[str]:
    model = _model_script(folder, meta)
    if isinstance(model, Err):
        return model
    sandbox = _sandbox(folder, meta)
    if isinstance(sandbox, Err):
        return sandbox
    stubs = _optional(folder, meta.stub_script)
    if isinstance(stubs, Err):
        return stubs
    if meta.extension_script is None:
        return Ok(_Scripts(model.value, sandbox.value, stubs.value, None))
    extensions = _model(_read(folder, "extensions.json"), ExtensionScript, "extensions.json")
    if isinstance(extensions, Err):
        return extensions
    return Ok(_Scripts(model.value, sandbox.value, stubs.value, extensions.value))


def read_case(root: Path, name: str) -> Ok[CaseDir] | Err[str]:
    folder = root / name
    meta = _model(_read(folder, "case.json"), CaseFile, "case.json")
    if isinstance(meta, Err):
        return meta
    log = folder / _own(folder, "log.jsonl")
    if not log.exists():
        return Err("log.jsonl is missing")
    scripts = _scripts(folder, meta.value)
    if isinstance(scripts, Err):
        return scripts
    appended = _appended(_read(folder, _own(folder, "expected.json")))
    if isinstance(appended, Err):
        return appended
    arts = sorted((folder / "artifacts").glob("*")) if (folder / "artifacts").exists() else []
    line0 = folder / "line0.json"
    s = scripts.value
    return Ok(
        CaseDir(
            name,
            meta.value,
            log.read_bytes(),
            tuple(p.read_bytes() for p in arts),
            s.model,
            s.sandbox,
            s.stubs,
            s.extensions,
            line0.read_bytes() if line0.exists() else None,
            appended.value[0],
            appended.value[1],
        )
    )


def case_names(root: Path) -> tuple[str, ...]:
    """Every case directory under `root`, by name."""
    return tuple(sorted(d.name for d in root.iterdir() if (d / "case.json").exists()))
